src/main/java/org/jenkinsci/plugins/github/webhook/subscriber/DuplicateEventsSubscriber.java

```diff
@@ -0,0 +1,135 @@
+     * @see com.cloudbees.jenkins.GitHubWebHook#doIndex
+     */
+    @Override
+    protected boolean isApplicable(@Nullable Item item) {
+        return false;
+    }
+
+    /**
+     * Subscribe to events that can trigger some kind of action within Jenkins, such as repository scan, build launch,
+     * etc.
+     * <p>
+     * There are about 63 specific events mentioned in the {@link GHEvent} enum, but not all of them are useful in
+     * Jenkins. Subscribing to and tracking them in duplicates tracker would cause an increase in memory usage, and
+     * those events' occurrences are likely larger than those that cause an action in Jenkins.
+     * <p>
+     * <a href="https://docs.github.com/en/webhooks/webhook-events-and-payloads">
+     *     Documentation reference (as also referenced in {@link GHEvent})</a>
+     * */
+    @Override
+    protected Set<GHEvent> events() {
+        return immutableEnumSet(
+                GHEvent.CHECK_RUN, // associated with GitHub action Re-run button to trigger build
+                GHEvent.CHECK_SUITE, // associated with GitHub action Re-run button to trigger build
+                GHEvent.CREATE, // branch or tag creation
+                GHEvent.DELETE, // branch or tag deletion
+                GHEvent.PULL_REQUEST, // PR creation (also PR close or merge)
+                GHEvent.PUSH // commit push
+            );
+    }
+
+    @Override
+    protected void onEvent(final GHSubscriberEvent event) {
+        String eventGuid = event.getEventGuid();
+        if (eventGuid == null) {
+            return;
+        }
+        long now = Instant.now().toEpochMilli();
+        EVENT_COUNTS_TRACKER.compute(
+                eventGuid, (key, value) -> new EventCountWithTTL(value == null ? 1 : value.count() + 1, now));
+    }
+
+    public static Map<String, Integer> getDuplicateEventCounts() {
+        return EVENT_COUNTS_TRACKER.entrySet().stream()
+                .filter(entry -> entry.getValue().count() > 1)
+                .collect(Collectors.toMap(
+                        Map.Entry::getKey, entry -> entry.getValue().count()));
+    }
+
+    @VisibleForTesting
+    static void cleanUpOldEntries(long now) {
+        EVENT_COUNTS_TRACKER
+                .entrySet()
+                .removeIf(entry -> (now - entry.getValue().lastUpdated()) > TTL_MILLIS);
+    }
+
+    /**
+     * Only for testing purpose
+     */
+    @VisibleForTesting
+    static Map<String, EventCountWithTTL> getEventCountsTracker() {
+        return new ConcurrentHashMap<>(EVENT_COUNTS_TRACKER);
+    }
+
+    @SuppressWarnings("unused")
+    @Extension
+    public static class EventCountTrackerCleanup extends PeriodicWork {
+
+        /**
+         * At present, as the {@link #TTL_MILLIS} is set to 10 minutes, we consider half of it for cleanup.
+         * This recurrence period is chosen to balance removing stale entries from accumulating in memory vs.
+         * additional load on Jenkins due to a new periodic job execution.
```

Comment:
[Caffeine](https://github.com/jenkinsci/caffeine-api-plugin) could be considered if you are looking for stronger assertions about the number of entries.

Reply:
Never knew about caffeine (checked now, it's good, thank you). But not sure if it is okay to bring in a new dependency into the plugin though.

The map here (`Map<String, Long>`) approximately taking 44 bytes per entry (String: storing GUID 36 bytes, Long: 8 bytes); with 1MB memory usage ~23k entries can be stored; this is seems like a sufficient for 10 minutes worth events from GH (for one controller) (?).

So thinking maybe bringing in caffeine would be adding another dependency in here.

Reply:
> a new dependency

https://plugins.jenkins.io/caffeine-api/ already used as a dep in various plugins including commonly installed ones, so no big deal I think, if it would make the code simpler / better.

Reply:
ok noted about it. currently changed the implementation to caffeine cache. ty